### ISSUE TITLE
Versioning scheme not semantic

### ISSUE THREAD

**Issue by mbirth** [google/closure-compiler-js#55]
The current versioning scheme not only breaks Composer (using the fxp/composer-asset-plugin to pull in dependencies from npm and bower) but also it is not [semantic versioning](https://docs.npmjs.com/getting-started/semantic-versioning) compliant. Is there any chance to change this in the future?

**Comment by concavelenz** [comment]
In what way does it break composer?

**Comment by mbirth** [comment]
The guys from Composer believe, (semantic) version numbers shouldn't have more than 6 digits per group (X.Y.Z being 3 groups). So `20170204.0.0` causes an error there and they believe it isn't their fault. (See the issue I linked in my opening comment.)

**Comment by Dominator008** [comment]
@mbirth I don't think Semantic Versioning restricts the number of digits in major version numbers.

**Comment by mbirth** [comment]
@Dominator008 Tell that those guys from Composer. (However, technically, major version bumps like you have with the dates would normally mean backwards incompatible API changes.)

**Comment by curry684** [comment]
Technically Semver does not impose an upper limit. It does however explicitly specify that the first public release of a package is to be called 1.0.0 and that a major version increase is reserved for breaking changes. Composer's limit is semi arbitrary of course but not more than MAXINT - both imply wrong usage of Semver. And if it were relaxed to 8 digits someone would complain because he wanted to have a full timestamp in there.

Both Composer and NPM use Semver, and your package is not. Don't blame the semver tools for breaking on wrong usage.

**Comment by curry684** [comment]
Let me clarify some more. First of all I'm not a member of the Composer team so my opinions are my own. There are also facts though: Composer has no intention whatsoever of handling NPM packages and repositories. The ecospheres are fundamentally incompatible after all.

A (imho horrible) plugin exists which tries to shoehorn the 2 worlds together, to save end users from maintaining 2 tools to manage 2 completely different tasks. It causes issues all the time because it bypasses Composer's APIs and abuses internal classes, causing breakages and frustration frequently. The issue at hand here is another prime example of that - there are no packages in the Composer ecosphere that can ever run into this limitation because it was there from day 1.

So the real solution to the problem is to use Composer for Composer packages, and NPM for NPM packages. We've wasted enough time of several peope now on an issue that wouldn't exist if all tools and packages would be used as intended. If anything, the FXP people should solve this issue, their plugin is causing it. Not Composer and not the Google Closure Compiler.

**Comment by concavelenz** [comment]
It is a pretty reasonable assumption that something in one of the compiler
options is a breaking change in every release.  While for most releases
upgrades are painless, compiler upgrades almost always break something.
Does it really matter to you if the version goes from 150.0.0 to 151.0.0 or
20170101.0.0 to 20170201.0.0 ?

**Comment by aluanhaddad** [comment]
Yeah, but date based versioning schemes are awkward for other reasons. They imply a point in time, not a state of being, if you'll pardon a bit of philosophical silliness, and some package managers _are_ designed to pull npm packages based on their semver tags. This would help anyone wanting to pull your package in via jspm for example.

I am not trying to be proscriptive. I just find that following common conventions really helps everyone in this domain.

Having said that, if semver does not make sense for your package, then you should not use it. Following conventions that are in opposition to your design, overall goals, or principles would be worse.

**Comment by teppeis** [comment]
In my understanding, Closure Library's date based versioning is semver compliant.
Because it actually has some breaking changes every time!

**Comment by curry684** [comment]
@teppeis as I stated before: *"Please refer to points 5, 8 and 10 of http://semver.org"* While we could argue the exact meaning of 'incremented' in point 8 the general consensus in the programming world is that something like the "increment operator" increments by 1 unless specified otherwise. And starts at 1.0.0. And Semver explicitly defines that build metadata, like for example the date of build, should be specified after the plus sign and not taken into account for version comparisons.

I'm not arguing that the Closure maintainers shouldn't do whatever they want, it's their choice and only causes problems when intermixing Composer and NPM via a monstrous aberration. But it's not Semver *compliant*, just sort of compatible depending on the library used for resolution.
